bugs/sru-19.4.33-manual-1.txt

```diff
@@ -0,0 +1,338 @@
+EOF
+
+cat > setup_proposed.sh <<EOF
+#/bin/bash
+mirror=http://archive.ubuntu.com/ubuntu
+echo deb \$mirror \$(lsb_release -sc)-proposed main | tee /etc/apt/sources.list.d/proposed.list
+apt-get update -q
+apt-get install -qy cloud-init
+EOF
+
+cat > byteswap_test.sh <<EOF
+#!/usr/bin/python3
+
+from cloudinit.stages import _pkl_load, _pkl_store
+
+import textwrap
+def swap_bytestring(s, width=2):
+        dd = [byte for byte in textwrap.wrap(s, 2)]
+        dd.reverse()
+        return ''.join(dd)
+
+def swapped_iid(iid):
+    parts = iid.split('-')
+    return '-'.join([
+        swap_bytestring(parts[0]),
+        swap_bytestring(parts[1]),
+        swap_bytestring(parts[2]),
+        parts[3],
+        parts[4]
+    ])
+
+# Cache a swapped instance-id so it looks like real product-uuid has swapped
+with open('/sys/class/dmi/id/product_uuid', 'r') as stream:
+    real_iid = stream.read().strip()
+ds =_pkl_load("/var/lib/cloud/instance/obj.pkl")
+fake_prev_iid = swapped_iid(ds.get_instance_id())
+# persist byte-swapped in /run/cloud-init/.instance-id as stages reads it
+with open('/var/lib/cloud/data/instance-id', 'w') as stream:
+    stream.write(fake_prev_iid)
+ds.metadata['instance-id'] = fake_prev_iid
+_pkl_store(ds, "/var/lib/cloud/instance/obj.pkl")
+
+assert ds.get_instance_id() == fake_prev_iid, "Did not swap cached instance-id"
+crawled_metadata = ds.crawl_metadata()
+assert crawled_metadata['metadata']['instance-id'] == fake_prev_iid, "Crawled metadata didn't return previous IID"
+print("SUCCESS")
+EOF
+chmod 755 byteswap_test.sh
+
+
+az group create --name $RESOURCE_GROUP --location $REGION
+
+sshopts="-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o LogLevel=ERROR"
+for series in $SRU_SERIES; do
+    echo "### BEGIN $series"
+    netcfg="/etc/netplan/50-cloud-init.yaml"
+    case $series in
+        xenial) image_version=16.04-DAILY-LTS
+                netcfg="/etc/network/interfaces.d/50-cloud-init.cfg";;
+        bionic) image_version=18.04-DAILY-LTS;;
+        disco) image_version=19.04-DAILY;;
+        eoan) image_version=19.10-DAILY;;
+        *) echo "!! UPDATE FAMILY CASE STATEMENT !!"; exit 1;;
+    esac
+    # Note: Gen2 instance types (--size) exhibit byte-swapped values
+    name=test-sru-$series
+    az vm create --name=$name \
+        --image=Canonical:UbuntuServer:$image_version:latest \
+        --admin-username=root --admin-username=ubuntu \
+        -g $RESOURCE_GROUP --ssh-key-value $SSH_KEY \
+        --custom-data sethostname.yaml --size Standard_B1s
+    IP=$(az vm list-ip-addresses --name $name | jq -r '.[] | .virtualMachine.network.publicIpAddresses[].ipAddress'| awk '{printf "ubuntu@%s", $1}')
+
+    echo "Created $name: ubuntu@$IP"
+
+    ssh $sshopts $IP -- cloud-init status --wait --long
+    scp $sshopts setup_proposed.sh byteswap_test.sh $IP:.
+    echo '--- Expect assert byteswap error for current release'
+    ssh $sshopts $IP -- sudo ./byteswap_test.sh
```

Comment:
Instead of examining internal state as part of `byteswap_test`, could we instead swap the bytes and reboot to confirm that we do see the incorrect new instance detection?